src/Sparrow.Server/Allocators.PointerTypes.cs

```diff
@@ -0,0 +1,527 @@
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new ReadOnlySpan<byte>((byte*)Address + start, length);
+        }
+
+        public static implicit operator Pointer(Pointer<byte> ptr)
+        {
+            return new Pointer(ptr.Address, ptr.SizeAsBytes);
+        }
+
+        public string Describe()
+        {
+            if (this.IsValid)
+                return $"{{{(long)this.Address:X32}|{this.Size}}}";
+
+            return "{null}";
+        }
+
+        public override string ToString()
+        {
+            return Describe();
+        }
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+#if !VALIDATE
+    // We do this to have the ability to perform extensive validation for pointers.
+    readonly
+#endif
+    public unsafe struct Pointer<T> : Meta.IDescribe, IPointerType<T> where T : struct
+    {
+        public readonly void* Address;
+
+        private readonly int _size;
+
+#if VALIDATE
+        public uint Generation;        
+#endif
+
+        [MethodImpl(MethodImplOptions.NoOptimization)]
+        static Pointer()
+        {
+            // If T is not a blittable type, fail on type loading unrecoverably.
+            // We need to ensure the JIT wont optimize away the call to Marshal.SizeOf<T> for correctness either now or in the future. 
+            Marshal.SizeOf<T>();
+        }
+
+        public Pointer(Pointer ptr)
+        {
+            this.Address = ptr.Address;
+            this._size = ptr.Size / Unsafe.SizeOf<T>();
+
+#if VALIDATE
+            this.Generation = ptr.Generation;
+#endif
+        }
+
+        public Pointer(void* ptr, int size)
+        {
+            this.Address = ptr;
+            this._size = size;
+
+#if VALIDATE
+            this.Generation = 0;
+#endif
+        }
+
+        public bool IsValid
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => Address != null;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> AsSpan()
+        {
+            return new Span<T>(Address, (int)Size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> AsSpan(int length)
+        {
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new Span<T>(Address, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> AsSpan(int start, int length)
+        {
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new Span<T>((byte*)Address + start * Unsafe.SizeOf<T>(), length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan()
```

Comment:
Pretty sure you meant `ReadOnlySpan<T>` here, no?

Reply:
Yes it is.